tfhe/src/shortint/engine/server_side/mod.rs

```diff
@@ -290,6 +290,20 @@ impl ShortintEngine {
         })
     }
 
+    pub(crate) fn generate_msg_lookup_table<F>(
+        &mut self,
+        server_key: &ServerKey,
+        f: F,
+        modulus: MessageModulus,
+    ) -> EngineResult<LookupTableOwned>
+    where
+        F: Fn(u64) -> u64,
+    {
+        Self::generate_lookup_table_with_engine(server_key, |x| {
+            f(x % modulus.0 as u64) % modulus.0 as u64
+        })
```

Comment:
I think I disagree with that % modulus at the end, someone may want to have bits in the carries even if only applying a LUT on the message part of x

Reply:
See https://github.com/zama-ai/tfhe-rs/pull/538#discussion_r1334440315

IMO, if someone wants to play with the input or output carry, they should use `generate_lookup_table` and handle the input and output carry themselves

Reply:
that's a design question, what is the meaning of the various functions we have in there, could you create a quick list of those function and their current behavior ? this way we can discuss and then agree on what we want, as those are mostly used by us (given shortint is already fairly low level) it's mainly a QoL for us so we can chose what we prefer as implementers

Reply:
- `generate_lookup_table` does not deal with modulus
- `generate_msg_lookup_table` applies `% msg_modulus` to the input and to the output of the function
- `generate_lookup_table_bivariate_with_factor`  which wrap `f` this way
```
let wrapped_f = |input: u64| -> u64 {
    let lhs = (input / factor_u64) % message_modulus;
    let rhs = (input % factor_u64) % message_modulus;

    f(lhs, rhs)
};
```
- `generate_lookup_table_bivariate` same as `generate_lookup_table_bivariate_with_factor` with `factor = server_key.message_modulus`

Reply:
If I take the integers implementations, most of them actually play with the carry space whether its the one of the input or output, I even have a lut that ouputs a block with padding bit set.

So as long as we have `generate_lookup_table` for maximum control and `generate_lookup_table_bivariate` its fine. 

`generate_msg_lookup_table` could be used in some places to replace `generate_lookup_table` where we manually apply msg_mod to both input/output but i don't think that is very common

Reply:
all right thanks for the detailed infos, will do a last round of reviews